### DISCUSSION TITLE
Change CPU frequency programatically on ESP82xx

### DISCUSSION THREAD

**Discussion by protectivedad** [arendst/Tasmota#21990] [Ideas]
I am connecting to a Mosquitto broker using TLS and a device certificate. They are light bulbs that have esp82xx chips in them. The connection time is about 1.2 seconds which is okay but noticeable. It drops in half if I use: `board_build.f_cpu = 160000000L`. Which is a lot better. I am worried about the bulb overheating and the wifi deteriorating. My ideal would be to turn down the frequency after the tls connects using the `system_update_cpu_freq` function.  Is this doable? Is this a useful feature to request?

**Comment by s-hadinger** [comment]
No. It's a compile time option 

That would be nice but it's not an option proposed by Espressif. 

I run all my esp8266 at 160MHz with no issue

**Reply by protectivedad** [reply]
> That would be nice but it's not an option proposed by Espressif.

If you look at the Espressif API here https://www.espressif.com/sites/default/files/2c-esp8266_non_os_sdk_api_guide_en_v1.5.4.pdf it says the function `system_update_cpu_freq` allows the changing of the running frequency. The function is in the firmware.map file from my generated firmwares:
`firmware.map:                0x0000000040287958                system_update_cpu_freq`

> I run all my esp8266 at 160MHz with no issue

Thanks. When I googled it there were comments on running 160 made the wifi unstable. Obviously that has not been your experience. I'll try it on one bulb and see how it works for me.

**Reply by s-hadinger** [reply]
Ok. I didn't know this api. But anyways changing frequency on the fly is not supported by Arduino core.

This feature would have so many impacts in timing sensitive code that anyways it will not be included, as it will create more problems than it solves

**Comment by sfromis** [comment]
If your bulb overheats, it would not be due to CPU frequency 160 MHz, but due the badness of the bulb form factor, with very limited space for the circuit to powering the leds. This can generate quite some heat if you go to the higher part of the brightness scale.

**Comment by protectivedad** [comment]
Thanks for the input. I have only one device (a bulb) that is powered by another device where I have to wait for it to connect before controlling it remotely. So It's not a big use-case :). From the perspective of maximizing performance while minimizing power consumption I thought it would be nice to turn up the processing power only when needed. If the 160 MHz is stable then there isn't really a problem just leaving it at the high speed. I'll leave this discussion open for a bit, I'd like to know if anyone has had any problems with running the esp82xx at 160 MHz?

**Reply by sfromis** [reply]
I've only got 4 ESP8266 with `"CpuFrequency":160`, but no trouble with them.
